File: src/telecom/repositories/postgres-subscriber-repository.ts
import { loadConfig, type ParsedEnvConfig } from '../../config/env.js';
import { getPool } from '../../persistence/pg-pool.js';
import type { TelecomSubscriber } from '../entities/telecom-subscriber.js';
import { buildFindByCellIdsSql, buildFindByMsisdnsSql } from './sql-builders.js';
import {
  type FindByCellIdsOptions,
  SubscriberRepositoryError,
  type SubscriberRepository,
  type SubscriberRow,
} from './subscriber-repository.js';

/**
 * Postgres subscriber repository — the sim's 1K → 300M path.
 *
 * Reads are parameterized and chunked by `SUBSCRIBER_LOOKUP_CHUNK_SIZE`.
 * Table + column names come from the SUBSCRIBER_COL_* env vars, so pointing
 * this at the real C-DOT subscriber schema is a .env change only.
 *
 * Writes (seeding) are handled by the seeder, which owns the sim's schema.
 */
export class PostgresSubscriberRepository implements SubscriberRepository {
  readonly name = 'telecom-sim:postgres';

  constructor(private readonly cfg: ParsedEnvConfig = loadConfig()) {}

  private columns() {
    const c = this.cfg;
    return {
      id: 'id',
      imsi: c.SUBSCRIBER_COL_IMSI,
      msisdn: c.SUBSCRIBER_COL_MSISDN,
      cellId: c.SUBSCRIBER_COL_CELL_ID,
      towerId: c.SUBSCRIBER_COL_TOWER_ID,
      technology: c.SUBSCRIBER_COL_TECHNOLOGY,
      status: c.SUBSCRIBER_COL_STATUS,
      lastSeen: c.SUBSCRIBER_COL_LAST_SEEN,
    };
  }

  async count(): Promise<number> {
    const pool = getPool();
    const result = await pool.query<{ n: string }>(`SELECT count(*)::text AS n FROM ${this.cfg.SUBSCRIBER_TABLE}`);
    return Number(result.rows[0]?.n ?? 0);
  }

  async upsertSubscribers(rows: readonly TelecomSubscriber[]): Promise<number> {
    throw new SubscriberRepositoryError('PostgresSubscriberRepository is read-only; use the seeder to load sim data');
  }

  async findByCellIds(cellIds: readonly string[], options: FindByCellIdsOptions = {}): Promise<SubscriberRow[]> {
    const pool = getPool();
    const chunkSize = this.cfg.SUBSCRIBER_LOOKUP_CHUNK_SIZE;
    const limit = options.limit ?? Number.POSITIVE_INFINITY;
    const cols = this.columns();
    const out: SubscriberRow[] = [];

    for (let i = 0; i < cellIds.length; i += chunkSize) {
      const chunk = cellIds.slice(i, i + chunkSize);
      const remaining = limit === Number.POSITIVE_INFINITY ? Number.POSITIVE_INFINITY : limit - out.length;
      if (remaining <= 0) break;
      const sql = buildFindByCellIdsSql(this.cfg.SUBSCRIBER_TABLE, cols, chunk, toSqlLimit(remaining, chunkSize));
      const result = await pool.query<{
        id: string;
        imsi: string;
        msisdn: string;
        cell_id: string;
        tower_id: string | null;
        technology: string | null;
        status: string | null;
        last_seen: string | Date | null;
      }>(sql.text, sql.values);
      for (const r of result.rows) {
        out.push({
          imsi: r.imsi,
          msisdn: r.msisdn,
          cellId: r.cell_id,
          towerId: r.tower_id ?? undefined,
          technology: r.technology ?? undefined,
          status: r.status ?? undefined,
          lastSeen: typeof r.last_seen === 'string' ? new Date(r.last_seen) : r.last_seen ?? undefined,
        });
      }
    }
    return out;
  }

  async findByMsisdns(msisdns: readonly string[], options: FindByCellIdsOptions = {}): Promise<TelecomSubscriber[]> {
    const pool = getPool();
    const limit = options.limit ?? Number.POSITIVE_INFINITY;
    const out: TelecomSubscriber[] = [];
    for (let i = 0; i < msisdns.length; i += this.cfg.SUBSCRIBER_LOOKUP_CHUNK_SIZE) {
      const chunk = msisdns.slice(i, i + this.cfg.SUBSCRIBER_LOOKUP_CHUNK_SIZE);
      const remaining = limit === Number.POSITIVE_INFINITY ? Number.POSITIVE_INFINITY : limit - out.length;
      if (remaining <= 0) break;
      const sql = buildFindByMsisdnsSql(this.cfg.SUBSCRIBER_TABLE, { msisdn: this.cfg.SUBSCRIBER_COL_MSISDN }, chunk, toSqlLimit(remaining, chunk.length));
      const result = await pool.query<TelecomSubscriber>(sql.text, sql.values);
      out.push(...result.rows);
    }
    return out;
  }
}

/** Resolve a JS limit to a SQL LIMIT (never above the chunk size). */
function toSqlLimit(jsLimit: number, chunkSize: number): number {
  return Number.isFinite(jsLimit) ? Math.min(jsLimit, chunkSize) : chunkSize;
}

File: src/main/java/com/turant/subscriber/PostgresSubscriberRepository.java
package com.turant.subscriber;

import com.turant.types.subscriber.Subscriber;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Repository;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

/**
 * PostgreSQL subscriber repository - the sim's 1K → 300M path.
 * 
 * Reads are parameterized and chunked by SUBSCRIBER_LOOKUP_CHUNK_SIZE.
 * Table + column names come from SUBSCRIBER_COL_* env vars, so pointing
 * this at the real C-DOT subscriber schema is a .env change only.
 * 
 * Migrated from TypeScript telecom/repositories/postgres-subscriber-repository.ts
 */
@Repository
public class PostgresSubscriberRepository implements SubscriberRepository {
    
    private static final Logger logger = LoggerFactory.getLogger(PostgresSubscriberRepository.class);
    
    private final JdbcTemplate jdbcTemplate;
    
    @Value("${subscriber.table:subscribers}")
    private String subscriberTable;
    
    @Value("${subscriber.column.imsi:imsi}")
    private String colImsi;
    
    @Value("${subscriber.column.msisdn:msisdn}")
    private String colMsisdn;
    
    @Value("${subscriber.column.cell-id:serving_cell_id}")
    private String colCellId;
    
    @Value("${subscriber.column.tower-id:tower_id}")
    private String colTowerId;
    
    @Value("${subscriber.column.technology:technology}")
    private String colTechnology;
    
    @Value("${subscriber.column.status:status}")
    private String colStatus;
    
    @Value("${subscriber.column.last-seen:last_seen}")
    private String colLastSeen;
    
    @Value("${subscriber.lookup-chunk-size:1000}")
    private int chunkSize;
    
    public PostgresSubscriberRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }
    
    @Override
    public String getName() {
        return "telecom-sim:postgres";
    }
    
    @Override
    public CompletableFuture<Long> count() {
        return CompletableFuture.supplyAsync(() -> {
            Long count = jdbcTemplate.queryForObject(
                "SELECT COUNT(*) FROM " + subscriberTable,
                Long.class
            );
            return count != null ? count : 0L;
        });
    }
    
    @Override
    public CompletableFuture<List<SubscriberRow>> findByCellIds(
            List<String> cellIds, 
            FindByCellIdsOptions options) {
        
        return CompletableFuture.supplyAsync(() -> {
            List<SubscriberRow> result = new ArrayList<>();
            int limit = options.getLimit() != null ? options.getLimit() : Integer.MAX_VALUE;
            
            // Process in chunks to avoid huge IN clauses
            for (int i = 0; i < cellIds.size(); i += chunkSize) {
                if (result.size() >= limit) {
                    break;
                }
                
                int end = Math.min(i + chunkSize, cellIds.size());
                List<String> chunk = cellIds.subList(i, end);
                
                int remaining = limit - result.size();
                int chunkLimit = Math.min(remaining, chunkSize);
                
                String sql = buildFindByCellIdsSql(chunk, chunkLimit);
                Object[] params = chunk.toArray();
                
                List<SubscriberRow> chunkResult = jdbcTemplate.query(sql, params, this::mapSubscriberRow);
                result.addAll(chunkResult);
            }
            
            logger.info("Found {} subscribers for {} cell IDs", result.size(), cellIds.size());
            return result;
        });
    }
    
    @Override
    public CompletableFuture<List<Subscriber>> findByMsisdns(
            List<String> msisdns, 
            FindByCellIdsOptions options) {
        
        return CompletableFuture.supplyAsync(() -> {
            List<Subscriber> result = new ArrayList<>();
            int limit = options.getLimit() != null ? options.getLimit() : Integer.MAX_VALUE;
            
            for (int i = 0; i < msisdns.size(); i += chunkSize) {
                if (result.size() >= limit) {
                    break;
                }
                
                int end = Math.min(i + chunkSize, msisdns.size());
                List<String> chunk = msisdns.subList(i, end);
                
                int remaining = limit - result.size();
                int chunkLimit = Math.min(remaining, chunkSize);
                
                String sql = buildFindByMsisdnsSql(chunk, chunkLimit);
                Object[] params = chunk.toArray();
                
                List<Subscriber> chunkResult = jdbcTemplate.query(sql, params, this::mapSubscriber);
                result.addAll(chunkResult);
            }
            
            return result;
        });
    }
    
    /**
     * Build SQL for finding subscribers by cell IDs.
     */
    private String buildFindByCellIdsSql(List<String> cellIds, int limit) {
        String placeholders = cellIds.stream()
            .map(id -> "?")
            .collect(Collectors.joining(", "));
        
        return String.format("""
            SELECT %s AS imsi, %s AS msisdn, %s AS cell_id, 
                   %s AS tower_id, %s AS technology, %s AS status, %s AS last_seen
            FROM %s
            WHERE %s IN (%s)
            LIMIT %d
            """,
            colImsi, colMsisdn, colCellId,
            colTowerId, colTechnology, colStatus, colLastSeen,
            subscriberTable,
            colCellId, placeholders,
            limit
        );
    }
    
    /**
     * Build SQL for finding subscribers by MSISDNs.
     */
    private String buildFindByMsisdnsSql(List<String> msisdns, int limit) {
        String placeholders = msisdns.stream()
            .map(id -> "?")
            .collect(Collectors.joining(", "));
        
        return String.format("""
            SELECT *
            FROM %s
            WHERE %s IN (%s)
            LIMIT %d
            """,
            subscriberTable,
            colMsisdn, placeholders,
            limit
        );
    }
    
    /**
     * Map result set to SubscriberRow.
     */
    private SubscriberRow mapSubscriberRow(ResultSet rs, int rowNum) throws SQLException {
        String imsi = rs.getString("imsi");
        String msisdn = rs.getString("msisdn");
        String cellId = rs.getString("cell_id");
        String towerId = rs.getString("tower_id");
        String technology = rs.getString("technology");
        String status = rs.getString("status");
        
        Instant lastSeen = null;
        java.sql.Timestamp ts = rs.getTimestamp("last_seen");
        if (ts != null) {
            lastSeen = ts.toInstant();
        }
        
        return new SubscriberRow(imsi, msisdn, cellId, towerId, technology, status, lastSeen);
    }
    
    /**
     * Map result set to full Subscriber.
     */
    private Subscriber mapSubscriber(ResultSet rs, int rowNum) throws SQLException {
        String msisdn = rs.getString(colMsisdn);
        String towerId = rs.getString(colTowerId);
        String cellId = rs.getString(colCellId);
        
        // Extract LAC from cell ID if available (simplified)
        String locationAreaCode = null;
        
        return new Subscriber(msisdn, towerId, locationAreaCode, cellId);
    }
}
